### PR TITLE
fix(raft): handle exceptions on partition server init

### PR DESCRIPTION
**Pull request by megglos**
## Description

Previously any RuntimeException happening in RaftPartitionServer#initServer lead to a broken future chain during start which lead to a stale node without any logs on the actual exception occurred during init. Ultimately flying silently till [here](https://github.com/camunda/zeebe/blob/main/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/PartitionManagerStep.java#L42) bringing the startup to a halt.

With this change issues are transparent, see this [log](https://console.cloud.google.com/logs/query;cursorTimestamp=2022-09-22T11:20:44.904454673Z;query=resource.labels.namespace_name%3D%22medic-cw-37-de38e9e086-benchmark-mixed%22%0Aresource.labels.pod_name%3D%22medic-cw-37-de38e9e086-benchmark-mixed-zeebe-2%22%0A-resource.labels.container_name%3D%22debugger-9q4tw%22%0A-logName%3D%22projects%2Fzeebe-io%2Flogs%2Fevents%22%0Atimestamp%3D%222022-09-22T11:20:44.904454673Z%22%0AinsertId%3D%2238ntsbk0c2ikn344%22%0Atimestamp%3D%222022-09-22T11:20:44.904454673Z%22%0AinsertId%3D%2238ntsbk0c2ikn344%22;summaryFields=:false:32:beginning;timeRange=2022-09-22T10:20:44.905Z%2F2022-09-22T11:20:44.905Z?project=zeebe-io) from a pod created with this change. 

This bug was hiding the underlying issue a node not being able to start due to #10451 .

## Related issues

relates to #10451

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftPartitionServer.java [hunk 1]
```diff
@@ -101,46 +100,45 @@ public RaftPartitionServer(
...
+      final long bootstrapStartTime;
+      log.info("Starting server for partition {}", partition.id());
+      final long startTime = System.currentTimeMillis();
+      final CompletableFuture<RaftServer> serverOpenFuture;
+      if (partition.members().contains(localMemberId)) {
+        if (server != null && server.isRunning()) {
+          return CompletableFuture.completedFuture(null);
+        }
+        synchronized (this) {
           initServer();
```
Comment (megglos) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftPartitionServer.java:114]: Caused by: java.lang.IllegalStateException: Expected to find a snapshot at index >= log's first index 213419494, but found snapshot 213252958. A previous snapshot is most likely corrupted.
		at io.atomix.raft.impl.RaftContext.verifySnapshotLogConsistent(RaftContext.java:226) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
		at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:203) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
		at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:258) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
		at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:232) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
		at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:190) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
		at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:156) ~[zeebe-atomix-cluster-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
```

I went for wrapping all invocations into a try block to make sure a future is always returned

File: atomix/cluster/src/test/java/io/atomix/raft/partition/impl/RaftPartitionServerTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.jupiter.api.Test;
+
+public class RaftPartitionServerTest {
+
+  private final RaftPartition raftPartition = mock(RaftPartition.class);
+  private final RaftPartitionGroupConfig partitionGroupConfig =
+      mock(RaftPartitionGroupConfig.class);
+
+  @Test
+  public void testInitServerRuntimeExceptionReturnsExceptionalFuture() {
```
Comment (megglos) [inline atomix/cluster/src/test/java/io/atomix/raft/partition/impl/RaftPartitionServerTest.java:35]: this is not really pretty as it depends on much internals of the RaftPartitionServer setup but I couldn't find an easier straightforward way to test this behavior

File: broker/src/main/java/io/camunda/zeebe/broker/bootstrap/PartitionManagerStep.java [hunk 3]
```diff
@@ -36,72 +36,65 @@ void startupInternal(
...
-        () ->
-            partitionManager
-                .start()
-                .whenComplete(
-                    (ok, error) -> {
-                      if (error != null) {
-                        startupFuture.completeExceptionally(error);
-                        return;
-                      }
+    CompletableFuture.runAsync(partitionManager::start)
```
Comment (megglos) [inline broker/src/main/java/io/camunda/zeebe/broker/bootstrap/PartitionManagerStep.java:39]: generally this was prone to silently fail if start failed synchronously, thus moving the completion/error handler one layer up.

File: broker/src/main/java/io/camunda/zeebe/broker/bootstrap/PartitionManagerStep.java [hunk 4]
```diff
@@ -36,72 +37,67 @@ void startupInternal(
...
-        () ->
-            partitionManager
-                .start()
-                .whenComplete(
-                    (ok, error) -> {
-                      if (error != null) {
-                        startupFuture.completeExceptionally(error);
-                        return;
-                      }
+    CompletableFuture.supplyAsync(partitionManager::start)
```
Comment (megglos) [inline broker/src/main/java/io/camunda/zeebe/broker/bootstrap/PartitionManagerStep.java:40]: generally this was prone to silently fail if start failed synchronously, thus moving the completion/error handler one layer up.

**Review by deepthidevaki — APPROVED** [review]
:rocket: Thanks @megglos

**Comment by megglos** [comment]
bors merge

**Comment by megglos** [comment]
bors retry

**Comment by oleschoenburg** [comment]
@megglos Just fyi, this needs to be backported manually: https://github.com/camunda/zeebe/pull/10450#issuecomment-1258030341

**Pull request merged**